### PR TITLE
refactor: use cache for push in AMQP plugin

### PR DESCRIPTION
**Pull request by aneojgurhem**

### CODE REVIEW

File: Adaptors/Amqp/src/PushQueueStorage.cs [hunk 1]
```diff
@@ -96,37 +94,43 @@ public class PushQueueStorage : QueueStorage, IPushQueueStorage
...
+                                                                                                                                                      .ConfigureAwait(false)),
+                                                                                                                                    Guid.NewGuid()
+                                                                                                                                        .ToString(),
+                                                                                                                                    s),
+                                                                                                      (link,
+                                                                                                       _) => new ValueTask<bool>(!link.IsClosed &&
+                                                                                                                                 !link.Session.IsClosed)));
```
Comment (lemaitre-aneo) [inline Adaptors/Amqp/src/PushQueueStorage.cs:117]: This should most likely be a private method. It it a bit hard to read now.

File: Adaptors/Amqp/src/PushQueueStorage.cs [hunk 2]
```diff
@@ -96,37 +94,43 @@ public class PushQueueStorage : QueueStorage, IPushQueueStorage
...
-                                                                                 MessageId = Guid.NewGuid()
-                                                                                                 .ToString(),
-                                                                               },
-                                                                })
-                                                     .ConfigureAwait(false);
+                                         var address = $"{partitionId}###q{whichQueue}";
+
+                                         var pool = senders_.GetOrAdd(address,
+                                                                      s => new ObjectPool<SenderLink>(parallelismLimit_,
+                                                                                                      async token => new SenderLink(new Session(await ConnectionAmqp
```
Comment (lemaitre-aneo) [inline Adaptors/Amqp/src/PushQueueStorage.cs:109]: So each Link have its own Session. Is it something we want to keep?

File: Adaptors/QueueCommon/src/Amqp.cs [hunk 3]
```diff
@@ -36,4 +36,5 @@ public class Amqp
   public bool   AllowHostMismatch { get; set; }
   public int    MaxRetries        { get; set; }
   public int    LinkCredit        { get; set; }
+  public int    ParallelismLimit  { get; set; } = 200;
```
Comment (lemaitre-aneo) [inline Adaptors/QueueCommon/src/Amqp.cs:39]: That is very high. I don't think it is a good idea. You can put 0 to use the number of threads by default.

File: Adaptors/Amqp/src/PushQueueStorage.cs [hunk 4]
```diff
@@ -96,37 +94,43 @@ public class PushQueueStorage : QueueStorage, IPushQueueStorage
...
+                                                                                                                                                      .ConfigureAwait(false)),
+                                                                                                                                    Guid.NewGuid()
+                                                                                                                                        .ToString(),
+                                                                                                                                    s),
+                                                                                                      (link,
+                                                                                                       _) => new ValueTask<bool>(!link.IsClosed &&
+                                                                                                                                 !link.Session.IsClosed)));
+
+                                         await pool.WithInstanceAsync(sender => sender.SendAsync(new Message(Encoding.UTF8.GetBytes(msgData.TaskId))
```
Comment (lemaitre-aneo) [inline Adaptors/Amqp/src/PushQueueStorage.cs:119]: You did not pass the cancellation token to SendAsync.

**Review by lemaitre-aneo — DISMISSED** [review]

File: Adaptors/Amqp/src/PushQueueStorage.cs (hunk 2 above)
Comment (aneojgurhem) [reply Adaptors/Amqp/src/PushQueueStorage.cs:109]: for now yes

File: Adaptors/Amqp/src/PushQueueStorage.cs (hunk 4 above)
Comment (aneojgurhem) [reply Adaptors/Amqp/src/PushQueueStorage.cs:119]: It does not take cancellation tokens

**Review by lemaitre-aneo — APPROVED** [review]

**Pull request merged**